File: Java/src/irrgarten/Player.java
///////////////////////////////////////
// Álvaro Maldonado Medina     	     //
// 2D - D3                           //
// Fichero: irrgarten/Player.java    //
///////////////////////////////////////

package irrgarten;

import java.util.ArrayList;

public class Player extends LabyrinthCharacter{
    
    private static final int MAX_WEAPONS = 2, MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10, HITS2LOSE = 3;
    
    private char number;
    private int consecutiveHits = 0;
    
    private ArrayList <Weapon> weapons; 
    private ArrayList <Shield> shields;
    
    private ShieldCardDeck shieldCardDeck;
    private WeaponCardDeck weaponCardDeck;

    
    // Constructores de la clase
    
    public Player(char number, float intelligence, float strength){ 
        
        super("Player #"+number, intelligence, strength, INITIAL_HEALTH);
        this.number = number;
        weapons = new ArrayList<>();
        shields = new ArrayList<>();
        
        shieldCardDeck = new ShieldCardDeck();
        weaponCardDeck = new WeaponCardDeck();
        
    }
    
    public Player(Player other){    //constructor de copia
        
        super(other);    
        this.number = other.number;
        this.weapons = other.weapons; 
        this.shields = other.shields;
        
        this.shieldCardDeck = other.shieldCardDeck;
        this.weaponCardDeck = other.weaponCardDeck;
    }
    
    // Método resurrect()
    // Reinicia las listas de armas y escudos a 0, establece el nivel de salud
    // al valor dado como salud inicial, y resetea el contador de hits
    // consecutivos a 0.
    
    public void resurrect()
    {   
        weapons.clear();    // reseteo de armas
        shields.clear();    // reseteo de escudos
        super.setHealth(INITIAL_HEALTH);
        resetHits();
    } 
    
    // Método getRow()
    // Devuelve la fila en la que se encuentra el jugador
    
    @Override
    public int getRow()
    {
        return super.getRow();
    }
    
    // Método getCol()
    // Devuelve la columna en la que se encuentra el jugador
    
    @Override
    public int getCol()
    {
        return super.getCol();
    }
    
    // Método getNumber()
    // Devuelve el num-ID del jugador
    
    public char getNumber()
    {
        return this.number;
    }
    
    // Método setPos
    // Modifica y establece la posición (row, col) del jugador
    
    @Override
    public void setPos(int row, int col)
    {
        super.setPos(row, col);
    }
    
    public Directions move(Directions direction, ArrayList<Directions> validMoves)
    {   
        
        int size = validMoves.size();
        boolean contained = validMoves.contains(direction);
        
        if ((size > 0) && (!contained)){
            return validMoves.get(0);
        }
        else{
            return direction;
        }
    }
    
    @Override
    public boolean defend(float receivedAttack)
    {
        return manageHit(receivedAttack);
    }  
        
    public void receiveReward(){
        
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        
        for (int i = 0; i < wReward; i++){
            Weapon wnew = weaponCardDeck.nextCard();      //idea que NO funciona
            //Weapon wNew = newWeapon();
            this.receiveWeapon(wnew);
        }
        
        for (int i = 0; i < sReward; i++){
            Shield snew = shieldCardDeck.nextCard();      //idea que NO funciona
            //Shield sNew = newShield();
            receiveShields(snew);   
        }
        
        int extraHealth = Dice.healthReward();
        float playerHealth = getHealth(); 
        setHealth(playerHealth + extraHealth);
        
    }
    
    // Método toString()
    // Devuelve el estado actual detallado del jugador
    
    @Override
    public String toString()
    {
        System.out.println("=============================================");
        System.out.println("Armas del jugador #"+number+": "+weapons);
        System.out.println("Escudos del jugador #"+number+": "+shields+"\n");
        return super.toString();
        
    }
    
    // Método receiveWeapon()
    //
    
    public void receiveWeapon(Weapon w)
    {    
        for (int i = weapons.size()-1; i >= 0; i--){
            boolean discard = weapons.get(i).discard();
            if (discard)
                weapons.remove(weapons.get(i));
        }
        
        int size = weapons.size();
        if (size < MAX_WEAPONS){
            weapons.add(w);
        }
    }
    
    public void receiveShields(Shield s)
    {
        for(int i = shields.size()-1; i >= 0; i--){
            boolean discard = shields.get(i).discard();
            if(discard)
                shields.remove(shields.get(i));
        }
        
        int size = shields.size();
        if(size < MAX_SHIELDS)
            shields.add(s);
    }
    
    // Método newWeapon()
    // Llama a la clase Weapon, crea una nuueva instancia de arma, y la 
    // añade a la lista de armas que hemos creado anteriormente
    
    public Weapon newWeapon()
    {
        Weapon newWeapon = new Weapon(Dice.weaponPower(),Dice.usesLeft());
        weapons.add(newWeapon);
        return newWeapon;
    }
    
    // Método newShield()
    // Igual que el método newWeapon, pero llamando a la clase Shield
    
    public Shield newShield()
    {
        Shield newShield = new Shield(Dice.shieldPower(), Dice.usesLeft());
        shields.add(newShield);
        return newShield;
        
    }
    
    // Método sumWeapons()
    // Devuelve la suma total de todas las armas, llamando a sus respectivos
    // métodos attack()
    
    protected float sumWeapons()   // idea
    {
        float sum = 0f;
        
        for (int i = 0; i < weapons.size(); i++){
            sum += weapons.get(i).attack();
        }
        
        return sum;
                
    }
    
    // Método sumShields()
    // Devuelve la suma total de todos los escudos, llamando a sus métodos
    // protect()
    
    protected float sumShields() 
    {
        float sum = 0f;
        
        for (int i = 0; i < shields.size(); i++){
            sum += shields.get(i).protect();
        }
        
        return sum;
    }
    
    // Método attack()
    // Calcula y devuelve la suma de la fuerza del jugador y lo aportado por
    // sus armas
    
    @Override
    public float attack(){
        return super.getStrength() + sumWeapons();
    }
    
    // Método defensiveEnergy()
    // Calcula y devuelve la suma de la inteligencia del jugador y lo
    // aportado por la protección de sus escudos
    
    protected float defensiveEnergy()
    {
        return super.getIntelligence() + sumShields();
    }
    
    public boolean manageHit(float receivedAttack)
    {
        float defense = defensiveEnergy();
        if (defense < receivedAttack) {
            super.gotWounded();
            incConsecutiveHits();
        }
        else {
            resetHits();
        }
       
        if ((consecutiveHits == HITS2LOSE) || (super.dead() )) {
            resetHits();
            return true;
        }
        else {
            return false;
        }
    }
    
    // Método resetHits
    // Resetea el contador de hits consecutivos a cero
    
    public void resetHits(){
        this.consecutiveHits = 0;
    }
    
    // Método incConsecutiveHits
    // Incrementa de uno en uno el contador de hits consecutivos del jugador
    
    public void incConsecutiveHits(){
        this.consecutiveHits++;
    }
    
    
}

File: Ruby/Player.rb
#encoding:utf-8

####################################### 
# Álvaro Maldonado Medina             #      
# 2D- D3                              #
# Fichero: Irrgarten/Player.rb        # 
#######################################

module Irrgarten
    class Player

        require_relative 'Weapon.rb'
        require_relative 'Shield.rb'
        require_relative 'Directions.rb'
        require_relative 'Dice.rb'

        @@MAX_WEAPONS = 2
        @@MAX_SHIELDS = 3
        @@INITIAL_HEALTH = 10
        @@HITS2LOSE = 3

        @consecutive_hits = 0
        attr_accessor :name, :number, :intelligence, :strength, :health,
                      :row, :col, :consecutive_hits, :weapons, :shields


        def initialize(number, intelligence, strength)
            
            @number = number
            @intelligence = intelligence
            @strength = strength
            @name = "Player#"+number.to_s
            @health = @@INITIAL_HEALTH
            @consecutive_hits = 0
            @row = -1
            @col = -1


            @weapons = Array.new {Weapon}
            @shields = Array.new {Shield}

            w = Weapon.new(1,1)
            s = Shield.new(1,1)
            @weapons.append(w)
            @weapons.append(s)

        end

        def resurrect
            @weapons.clear
            @shields.clear
            @health = @@INITIAL_HEALTH
            reset_hits
        end

        def row
            @row
        end

        def col
            @col
        end

        def number
            @number
        end

        def set_pos(row, col)
            @row = row
            @col = col
        end

        def dead
            health <= 0
        end

        def move(direction, valid_moves)
            size = valid_moves.size
            contained = valid_moves.include?(direction)

            if size > 0 && !contained
                valid_moves[0]
            else
                direction
            end

        end

        def attack
            @strength + sum_weapons
        end

        def defend(received_attack)
            manage_hit(received_attack)
        end

        def receive_reward
            w_reward = Dice.weapons_reward
            s_reward = Dice.shields_reward

            w_reward.times do
                wnew = new_weapon
                receive_weapon(wnew)
            end

            s_reward.times do
                snew = new_shield
                receive_shields(snew)
            end

            extra_health = Dice.health_reward
            @health += extra_health
        end

        def to_string
            puts "#{name}, H: #{health}, I: #{intelligence}, S: #{strength},
            Pos: #{row},#{col}"
        end

        private
        def receive_weapon(w)
            (weapons.size - 1).downto(0) do |i|
                discard = weapons[i].discard
                weapons.delete_at(i) if discard
            end

            size = weapons.size
            weapons.push(w) if size < @@MAX_WEAPONS
        end

        def receive_shields(s)
            (shields.size - 1).downto(0) do |i|
                discard = shields[i].discard
                shields.delete_at(i) if discard
            end

            size = shields.size
            shields.push(s) if size < @@MAX_WEAPONS
        end

        def new_weapon
            Weapon(Dice.weapon_power, Dice.uses_left)
        end

        def new_shield
            Shield(Dice.shield_power, Dice.uses_left)
        end

        def sum_weapons

            sum = 0
            (0..weapons.length).each { |i|
                sum += weapons[i].attack.to_f
            }

            sum
        end

        def sum_shields

            sum = 0
            (0..shields.length).each { |i|
                sum += shields[i].protect
            }

            sum
        end

        def defensive_energy
            @intelligence + sum_shields
        end

        def manage_hit(received_attack)
            defense = defensive_energy

            if defense < received_attack
                got_wounded
                inc_consecutive_hits
            else
                reset_hits
            end


            if consecutive_hits == @@HITS2LOSE || dead
                reset_hits
                lose = true
            else
                lose = false
            end

            lose
        end

        def reset_hits
            @consecutive_hits = 0
        end

        def got_wounded
            @health -= 1
        end

        def inc_consecutive_hits
            @consecutive_hits += 1
        end
    end


end



